### DISCUSSION TITLE
Add the full input function

### DISCUSSION THREAD

**Discussion by L2750558108** [godotengine/godot-proposals#6022] [Scripting]
![image](https://user-images.githubusercontent.com/119792950/210044344-52d640d5-b980-4ffa-8709-4112105d2ddc.png)

Add 
is_key_button_just_pressed, 
is_key_button_just_released, 
is_joy_button_just_pressed, 
is_joy_button_just_released, 
is_mouse_button_just_pressed, 
is_mouse_button_ just_released, 
is_physical_key_just_pressed, 
is_physical_key_just_released
methods to the Input class.
Their functionality refers to is_action_just_pressed and is_action_just_released methods,used to check if a key is just pressed or released.

I think there should be a full input checking function for users who don't use is_action methods

Dont forget C#

**Comment by skeddles** [comment]
just ran into this, i dont understand why i can have is_action_just_pressed, which works for a single controller, but not is_joy_button_just_pressed so it can work for multiple controllers

**Reply by theraot** [reply]
You can have actions for each player, each configured to use a different controller.

---

Alternatively, I suppose you can keep track of what was pressed the prior frame, or handle the actions in `_input`.

---

To be clear, I'm not against having the functions mentioned by OP.

---

The way I currently handle this is by remapping actions in runtime. I have a helper function to remap actions, it is like this:

```GDScript
static func set_action_events(action_name:StringName, events:Array[InputEvent]) -> void:
	if action_name.is_empty() or not InputMap.has_action(action_name):
		return

	InputMap.action_erase_events(action_name)
	for new_event in events:
		InputMap.action_add_event(action_name, new_event)
```

So you can tell the user to press whatever key or button they want, you get it as an `InputEvent` in `_input` and map the action to it.

Of course, you want the game to save the settings and load them on start up, and also give the option to the player to reset the actions either to defaults or to what was last saved.

*I suppose this is too much work for a quick prototype, but it is saving you work in a well polished game compared to having to implement action remapping yourself. There is, of course, the concern of saving and loading resources with Godot methods introducing a way to inject code... And if you are making your own serializer (as I did), perhaps the work it saves is not that much.*

---

Anyway, I hope you can find a path forward for your project.
